{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c3a3ec9-ef67-4728-a50a-87c84506abcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### install Julia packages #####\n",
    "PkgList = [\"CSV\", \"DataFrames\", \"KrylovKit\", \"StatsBase\", \"LinearAlgebra\", \"Statistics\", \"MultivariateStats\", \"Distributions\", \"KernelDensity\", \"ParallelKMeans\", \"Clustering\", \"UMAP\", \"RCall\", \"NetworkLayout\", \"GraphPlot\", \"Graphs\", \"Colors\", \"Compose\", \"CairoMakie\", \"Fontconfig\", \"JLD2\"]\n",
    "using Pkg; for p in PkgList; Pkg.add(p); end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95640a55",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "##### install R packages #####\n",
    "install.packages(c(\"BiocManager\",\"ggplot2\",\"RColorBrewer\"), repos=\"https://ftp.yz.yamagata-u.ac.jp/pub/cran\")\n",
    "BiocManager::install(c(\"clusterProfiler\",\"enrichplot\",\"org.Hs.eg.db\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "124e39d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### SGCRNA #####\n",
    "module SGCRNAs\n",
    "    using Dates\n",
    "    using CSV, DataFrames\n",
    "    using KrylovKit\n",
    "    using StatsBase\n",
    "    using LinearAlgebra, Statistics, MultivariateStats, Distributions, KernelDensity\n",
    "    using ParallelKMeans, Clustering\n",
    "    using Random, UMAP\n",
    "    using RCall\n",
    "    using NetworkLayout, GraphPlot, Graphs, Colors\n",
    "    using Compose, CairoMakie, Fontconfig\n",
    "    using JLD2\n",
    "\n",
    "    ##### get rank of element #####\n",
    "        function ElmRank(arr::Vector, revflg::Bool)\n",
    "            sorted_indices = sortperm(arr, rev=revflg)\n",
    "            sorted_arr = arr[sorted_indices]\n",
    "            ranks = similar(arr)\n",
    "\n",
    "            rank = 1\n",
    "            for i in 1:length(arr)\n",
    "                if i > 1 && sorted_arr[i] != sorted_arr[i - 1]\n",
    "                    rank = i\n",
    "                end\n",
    "                ranks[sorted_indices[i]] = rank\n",
    "            end\n",
    "\n",
    "            return ranks\n",
    "        end\n",
    "    ##### get rank of element #####\n",
    "\n",
    "    ##### correlation matrix calculation #####\n",
    "        \"\"\"\n",
    "        argument\n",
    "        gene: gene name list\n",
    "        data: gene expression matrix\n",
    "        fn: prefix of save file name. defalut: do not save\n",
    "            output file name: fn_cor.tsv\n",
    "        threshold: value for remove genes with more than a certain number of zeros\n",
    "        mode: mode of measurement errors elimination\n",
    "                :NONE -> measurement error is not considered (Select this option when there are sufficient number of samples; Default)\n",
    "                :LESS -> defined as the value below the mode\n",
    "                :SIGMA -> defined as the value below the 2σ(mode is considered as σ)\n",
    "                :FTEST -> defined as the value below the significantly different from the mode by pval\n",
    "        binSize: histogram bin size used to determine measurement error when mode is other than :NONE\n",
    "        pval: p-value for determining measurement error using :FTEST & Statistical tests of correlation coefficients\n",
    "        power: Power in statistical tests of correlation coefficients.\n",
    "\n",
    "        returns\n",
    "        CorData: correlation matrix\n",
    "        GradData: gradient matrix\n",
    "        \"\"\"\n",
    "        function CCM(gene::Vector, data::Matrix; fn::String=\"\", threshold::Float64=0.5, mode::Symbol=:NONE, binSize::Float64=0.01, pval::Float64=0.05, power::Float64=0.8)\n",
    "            SmplNum = size(data, 2)\n",
    "        \n",
    "            # Remove genes with more than a certain number of zeros\n",
    "            Q = map(x -> sum(x .== 0.0), eachrow(data))\n",
    "            Gene = gene[Q .< SmplNum * threshold]\n",
    "            Data = data[Q .< SmplNum * threshold, :]\n",
    "        \n",
    "            # Eliminates measurement errors\n",
    "            if mode != :NONE\n",
    "                # Get the mode of the coefficient of variation\n",
    "                cvList = map(x -> std(x, corrected=false), eachrow(Data)) ./ map(mean, eachrow(Data))\n",
    "                freqCurve = pdf(kde(cvList), [0.0:binSize:ceil(maximum(cvList))])[1]\n",
    "                cvMode = binSize * (argmax(freqCurve) - 1)\n",
    "\n",
    "                # plot freqCurve\n",
    "                xs = collect(0.0:binSize:ceil(maximum(cvList)))\n",
    "                f = Figure(size = (1300, 400))\n",
    "                ax1 = Axis(f[1, 1], title = \":LESS\", ylabel=\"probability density\", xautolimitmargin=(0.0f0, 0.0f0), yautolimitmargin=(0.0f0, 0.0f0))\n",
    "                ax2 = Axis(f[1, 2], title = \":SIGMA\", xautolimitmargin=(0.0f0, 0.0f0), yautolimitmargin=(0.0f0, 0.0f0), yticklabelsvisible=false)\n",
    "                ax3 = Axis(f[1, 3], title = \":FTEST\", xautolimitmargin=(0.0f0, 0.0f0), yautolimitmargin=(0.0f0, 0.0f0), yticklabelsvisible=false)\n",
    "                linkyaxes!(ax1, ax2)\n",
    "                linkyaxes!(ax1, ax3)\n",
    "                hidespines!(ax1, :t, :r)\n",
    "                hidespines!(ax2, :t, :r)\n",
    "                hidespines!(ax3, :t, :r)\n",
    "                hidedecorations!(ax1, label=false, ticklabels=false, ticks=false)\n",
    "                hidedecorations!(ax2, label=false, ticklabels=false, ticks=false)\n",
    "                hidedecorations!(ax3, label=false, ticklabels=false, ticks=false)\n",
    "                barplot!(ax1, xs, freqCurve, color=[x <= cvMode ? \"red\" : \"blue\" for x in xs], gap=0)\n",
    "                barplot!(ax2, xs, freqCurve, color=[x <= cvMode*2 ? \"red\" : \"blue\" for x in xs], gap=0)\n",
    "                barplot!(ax3, xs, freqCurve, color=[x <= cvMode*sqrt(quantile(FDist(SmplNum-1, SmplNum-1), 1 - pval)) ? \"red\" : \"blue\" for x in xs], gap=0)\n",
    "                Label(f[end+1, :], text=\"standard deviation\")\n",
    "                CairoMakie.save(fn * \"_freqCurve.png\", f)\n",
    "\n",
    "                if mode == :LESS\n",
    "                    # Remove cvMode and below\n",
    "                elseif mode == :SIGMA\n",
    "                    # Assume cvMode as σ and remove 2σ\n",
    "                    cvMode *= 2\n",
    "                elseif mode == :FTEST\n",
    "                    # Delete cv that is below the significance level in F test\n",
    "                    cvMode *= sqrt(quantile(FDist(SmplNum-1, SmplNum-1), 1 - pval))\n",
    "                end\n",
    "        \n",
    "                Gene = Gene[cvList .> cvMode]\n",
    "                Data = Data[cvList .> cvMode, :]\n",
    "            end\n",
    "        \n",
    "            # Calculate covariance and standard deviation\n",
    "            Avg = map(mean, eachrow(Data))\n",
    "            Dist = Data .- Avg\n",
    "            Covar = (Dist * Dist')\n",
    "            Std = map(x -> sqrt(sum(x .^ 2)), eachrow(Dist))\n",
    "            # Calculate correlation coefficients in batches\n",
    "            CorMat = Covar ./ (Std * Std')\n",
    "            # Calculate gradients in batches\n",
    "            VarVec = map(x -> sum(x .^ 2), eachrow(Dist))\n",
    "            GeneNum = length(VarVec)\n",
    "            VarMat = Matrix{Float64}(undef, GeneNum, GeneNum)\n",
    "            for i in 1:GeneNum\n",
    "                VarMat[i, 1:i] .= VarVec[i]\n",
    "            end\n",
    "            VarMat += VarMat'\n",
    "            for i in 1:GeneNum\n",
    "                VarMat[i, i] -= VarVec[i]\n",
    "            end\n",
    "            GradMat = Covar ./ VarMat\n",
    "\n",
    "            # Set all but statistically significant correlation coefficients to zero\n",
    "            d = Normal()\n",
    "            Za = quantile(Normal(), 1-pval/2)\n",
    "            Zb = quantile(Normal(), power)\n",
    "            z = exp(2*(Za+Zb) / sqrt(SmplNum-3))\n",
    "            r = (z-1) / (z+1)\n",
    "            CorMat .*= (abs.(CorMat) .>= r)\n",
    "\n",
    "            # Conversion to data frame\n",
    "            CorData = hcat(\n",
    "                        DataFrame(Symbol=Gene),\n",
    "                        DataFrame(CorMat, Gene)\n",
    "                    )\n",
    "            CorData = coalesce.(CorData, 0.0)\n",
    "            replace!.(eachcol(CorData), NaN => 0.0)\n",
    "            GradData = hcat(\n",
    "                        DataFrame(Symbol=Gene),\n",
    "                        DataFrame(GradMat, Gene)\n",
    "                    )\n",
    "            GradData = coalesce.(GradData, 0.0)\n",
    "            replace!.(eachcol(GradData), NaN => 0.0)\n",
    "                \n",
    "            # Genes that did not correlate with any of the genes were removed\n",
    "            CorData = CorData[sum.(eachrow(CorData[:,2:end])) .!= 0.0, :]\n",
    "            GradData = innerjoin(CorData[:,[:Symbol]], GradData, on=:Symbol)\n",
    "\n",
    "            # alignment\n",
    "            sort!(CorData, :Symbol)\n",
    "            CorData = CorData[:, sort(CorData.Symbol)]\n",
    "            sort!(GradData, :Symbol)\n",
    "            GradData = GradData[:, sort(GradData.Symbol)]\n",
    "        \n",
    "            # Keep 0 between selves.\n",
    "            for i in 1:ncol(CorData)\n",
    "                CorData[i, i] = 0.0\n",
    "                GradData[i, i] = 0.0\n",
    "            end\n",
    "\n",
    "            save_object(fn * \"_cor.jld2\", CorData)\n",
    "            save_object(fn * \"_grad.jld2\", GradData)\n",
    "\n",
    "            return CorData, GradData\n",
    "        end\n",
    "        export CCM\n",
    "    ##### correlation matrix calculation #####\n",
    "\n",
    "    ##### Laplacian matrix calculation #####\n",
    "        function LaplacianMatrix(df::DataFrame, normFlg::Bool, randNormFlg::Bool)\n",
    "            mat = Matrix(df)\n",
    "            nodeScores = sum.(eachrow(mat))\n",
    "            matD = diagm(nodeScores)\n",
    "            matL = matD .- mat\n",
    "            if(normFlg)\n",
    "                matL = sqrt(inv(matD)) * matL * sqrt(inv(matD))\n",
    "            elseif(randNormFlg)\n",
    "                matL = inv(matD) * matL\n",
    "            end\n",
    "\n",
    "            return matL\n",
    "        end\n",
    "    ##### Laplacian matrix calculation #####\n",
    "\n",
    "    ##### clustering #####\n",
    "        function Clustering_Eigen(matL::Matrix, k::Int64, normFlg::Bool)\n",
    "            eigVals, eigVecs, eigInfo = eigsolve(matL, k+10, :SR, krylovdim=5*k)\n",
    "            eigVals = Real.(eigVals)\n",
    "            eigVecs = Real.(reduce(hcat, eigVecs))\n",
    "            eigVecs = eigVecs[:, 1e-8 .< eigVals]\n",
    "\n",
    "            embedding = eigVecs'[1:k, :]\n",
    "            if (normFlg)\n",
    "                buf = map(x -> x ./ sum(x .^ 2), eachrow(embedding))\n",
    "                embedding = reduce(hcat, buf)'\n",
    "            end\n",
    "            \n",
    "            return Matrix(embedding)\n",
    "        end\n",
    "        function Clustering_Main(emb::Matrix, seed::Int64, k::Int64, itr::Int64)\n",
    "            RndSeed = Random.seed!(seed)\n",
    "\n",
    "            embedding = deepcopy(emb[1:k, :])\n",
    "            # k-means clustering by automatic k-value determination\n",
    "            if size(embedding, 2) < 100\n",
    "                res = ParallelKMeans.kmeans(Hamerly(), embedding, k, max_iters=itr, rng=RndSeed)\n",
    "            elseif size(embedding, 2) > 10000\n",
    "                res = ParallelKMeans.kmeans(Elkan(), embedding, k, max_iters=itr, rng=RndSeed)\n",
    "            else\n",
    "                res = ParallelKMeans.kmeans(Yinyang(), embedding, k, max_iters=itr, rng=RndSeed)\n",
    "            end\n",
    "            \n",
    "\n",
    "            score = [0.0,0.0,0.0,0.0]\n",
    "            score[1] = clustering_quality(embedding, res.centers, res.assignments, quality_index=:xie_beni)\n",
    "            score[2] = clustering_quality(embedding, res.centers, res.assignments, quality_index = :davies_bouldin)\n",
    "            score[3] = clustering_quality(embedding, res.assignments, quality_index=:silhouettes)\n",
    "            score[4] = clustering_quality(embedding, res.centers, res.assignments, quality_index = :calinski_harabasz)\n",
    "            \n",
    "            return embedding, res, score\n",
    "        end\n",
    "    ##### clustering #####\n",
    "\n",
    "    ##### SpectralClustering #####\n",
    "        function SCSub(emb::Matrix, kmin::Int64, kstep::Int64, kmax::Int64, itr::Int64, seed::Int64)\n",
    "            resDf = DataFrame(k=[], assign=[], xbi=[], dbi=[], si=[], chi=[])\n",
    "            for k in kmin:kstep:kmax\n",
    "                embedding, res, score = Clustering_Main(emb, seed, k, itr)\n",
    "                push!(resDf, hcat([k], [res.assignments], score'))\n",
    "            end\n",
    "            resDf.k = float.(resDf.k)\n",
    "            resDf[!, 3:6] = float.(resDf[:, 3:6])\n",
    "\n",
    "            # auto detect value\n",
    "            RankList = [ElmRank(resDf.si,true),ElmRank(resDf.chi,true),ElmRank(resDf.xbi,false),ElmRank(resDf.dbi,false)]\n",
    "            RankList = hcat(RankList...)\n",
    "            Score = map(prod, eachrow(RankList))\n",
    "            minK = argmin(Score)\n",
    "\n",
    "            return resDf.assign[minK]\n",
    "        end\n",
    "        \"\"\"\n",
    "        argument\n",
    "        cor: dataframe of correlation matrix (return value of CCM())\n",
    "        grad: dataframe of gradient matrix (return value of CCM())\n",
    "        tNodeNum: threshold of sub-cluster node number; default: 100\n",
    "        depthMax: Depth of sub-clusters; default: 5\n",
    "        kmin,kstep,kmax: test range of clusters number; default: 3, 1, 15\n",
    "        itr: number of trials; default: 300\n",
    "        seed: seed value of random number; default: 42 (Answer to the Ultimate Question of Life, the Universe, and Everything)\n",
    "        nNeighbors: UMAP parameter; default: 40\n",
    "        minDist: UMAP parameter; default: 0.01\n",
    "        normFlg: Whether to symmetrically normalize the Laplacian matrix; default: false\n",
    "        randNormFlg: Whether to random walk normalize the Laplacian matrix; default: false\n",
    "\n",
    "        returns\n",
    "        clust: cluster number of each gene\n",
    "        pos: gene position for drawing network\n",
    "        edgeScore: edge score for drawing network\n",
    "        \"\"\"\n",
    "        function SpectralClustering(cor::DataFrame, grad::DataFrame; tNodeNum::Int64=100, depthMax::Int64=5, kmin::Int64=3, kstep::Int64=1, kmax::Int64=15, itr::Int64=300, seed::Int64=42, nNeighbors::Int64=40, minDist::Float64=0.01, normFlg::Bool=true, randNormFlg::Bool=false)\n",
    "            rowNum = size(cor, 1)\n",
    "            df = ((1 .+ cor) ./ 2) .* exp.(-1 .* abs.(log.(abs.(grad))))\n",
    "            # Laplacian matrix calculation\n",
    "            matL = LaplacianMatrix(df, normFlg, randNormFlg)\n",
    "            emb = Clustering_Eigen(matL, kmax, normFlg)\n",
    "            clust = SCSub(emb, kmin, kstep, kmax, itr, seed)\n",
    "            clustData = [clust]\n",
    "            kNum = maximum(clust[1]); d = 0;\n",
    "            while ((maximum(map(x -> sum(clustData[d+1] .== x), 1:kNum)) > tNodeNum) & (d < depthMax))\n",
    "                append!(clustData, deepcopy([clustData[d+1]]))\n",
    "                for k in 1:kNum\n",
    "                    subEmb = emb[:, clustData[d+2] .== k]\n",
    "                    if size(subEmb, 2) > tNodeNum\n",
    "                        subClust = SCSub(subEmb, kmin, kstep, kmax, itr, seed) .- 1\n",
    "                        subClust[subClust .!= 0] .+= kNum\n",
    "                        subClust[subClust .== 0] .= k\n",
    "                        clustData[d+2][clustData[d+2] .== k] = subClust\n",
    "                        kNum = maximum(subClust)\n",
    "                    end\n",
    "                end\n",
    "                d += 1\n",
    "            end\n",
    "\n",
    "            kNum = maximum(clustData[1])\n",
    "            embedding = umap(emb[1:kNum, :], 2; n_neighbors=nNeighbors, min_dist=minDist)\n",
    "\n",
    "            # # plot corrected SSE\n",
    "            # ks = collect(kmin:kstep:kmax)\n",
    "            # f = Figure(size = (1125, 750))\n",
    "            # ax1 = Axis(f[1, 1], title = \"Xie-Beni\")\n",
    "            # ax2 = Axis(f[1, 2], title = \"Davies-Bouldin\")\n",
    "            # ax3 = Axis(f[2, 1], title = \"silhouettes\")\n",
    "            # ax4 = Axis(f[2, 2], title = \"Calinski-Harabasz\")\n",
    "            # hidespines!(ax1, :t, :r)\n",
    "            # hidespines!(ax2, :t, :r)\n",
    "            # hidespines!(ax3, :t, :r)\n",
    "            # hidespines!(ax4, :t, :r)\n",
    "            # hidedecorations!(ax1, label=false, ticklabels=false, ticks=false)\n",
    "            # hidedecorations!(ax2, label=false, ticklabels=false, ticks=false)\n",
    "            # hidedecorations!(ax3, label=false, ticklabels=false, ticks=false)\n",
    "            # hidedecorations!(ax4, label=false, ticklabels=false, ticks=false)\n",
    "            # maxAe = []\n",
    "            # scatterlines!(ax1, resDf.k, resDf.xbi)\n",
    "            # scatterlines!(ax2, resDf.k, resDf.dbi)\n",
    "            # scatterlines!(ax3, resDf.k, resDf.si)\n",
    "            # scatterlines!(ax4, resDf.k, resDf.chi)\n",
    "\n",
    "            # Label(f[:, 0], text=\"Score\", rotation=pi/2)\n",
    "            # Label(f[end+1, :], text=\"k\")\n",
    "\n",
    "            # CairoMakie.save(fn, f)\n",
    "\n",
    "            return clustData, Matrix(permutedims(embedding)), cor .* exp.(-1 .* abs.(log.(abs.(grad))))\n",
    "        end\n",
    "        export SpectralClustering\n",
    "    ##### SpectralClustering #####\n",
    "\n",
    "    ##### draw network #####\n",
    "        \"\"\"\n",
    "        argument\n",
    "        df: dataframe of correlation matrix (return value of CCM())\n",
    "        clust: cluster number of each gene (one of return value of SpectralClustering())\n",
    "        pos: gene position for drawing network (one of return value of SpectralClustering())\n",
    "        il: module number list which you want to draw\n",
    "\n",
    "        return\n",
    "        nw: undirected graph\n",
    "        pos: node position\n",
    "        cnctdf: converted correlation matrix\n",
    "        clust: cluster number of each gene in network\n",
    "        score: node scores\n",
    "        \"\"\"    \n",
    "        function SetNetwork(df::DataFrame, clust::Vector{Int64}, pos::Matrix; il::Vector=[])\n",
    "            ##### preliminaries #####\n",
    "                # All clusters you want to draw if none are specified.\n",
    "                if length(il) == 0\n",
    "                    il = sort(unique(clust))\n",
    "                end\n",
    "                # Extract genes present in the cluster you want to draw\n",
    "                Q1 = (clust .== il[1])\n",
    "                if length(il) > 1\n",
    "                    for i in 2:length(il)\n",
    "                        Q1 .|= (clust .== il[i])\n",
    "                    end\n",
    "                end\n",
    "                cnctdf = deepcopy(Matrix(df[Q1, Q1]))\n",
    "                gene_list = names(df)[Q1]\n",
    "                Q2 = (map(sum, eachrow(cnctdf)) .!= 0.0)\n",
    "                cnctdf = cnctdf[Q2, Q2]\n",
    "                gene_list = gene_list[Q2]\n",
    "                gene_num = length(gene_list)\n",
    "\n",
    "                # Convert to upper triangular matrix\n",
    "                triu!(cnctdf)\n",
    "                cnctdf = DataFrame(hcat(gene_list,cnctdf), vcat([\"Symbol\"],gene_list))\n",
    "                cnctdf = stack(cnctdf, 2:ncol(cnctdf))\n",
    "                rename!(cnctdf, [:e1,:e2,:cor])\n",
    "                # Remove correlations in the same gene and duplicate combinations\n",
    "                ## Correlation coefficients between themselves are set to 0.\n",
    "                ## The overlapping combinations have a correlation coefficient of zero\n",
    "                ## due to the conversion to an upper triangular matrix.\n",
    "                cnctdf = cnctdf[cnctdf.cor .!= 0.0, :]\n",
    "                sort!(cnctdf, :e1)\n",
    "            \n",
    "                # Assign node numbers to genes\n",
    "                buf = DataFrame(e1=sort(unique(vcat(cnctdf.e1, cnctdf.e2))))\n",
    "                buf[!, :i1] = collect(1:nrow(buf))\n",
    "                cnctdf = innerjoin(cnctdf, buf, on=:e1)\n",
    "                rename!(buf, [:e2,:i2])\n",
    "                cnctdf = innerjoin(cnctdf, buf, on=:e2)\n",
    "\n",
    "                # Assign module numbers to genes\n",
    "                buf = DataFrame(e1=gene_list, m1=clust[Q1][Q2])\n",
    "                cnctdf = innerjoin(cnctdf, buf, on=:e1)\n",
    "                rename!(buf, [:e2,:m2])\n",
    "                cnctdf = innerjoin(cnctdf, buf, on=:e2)\n",
    "            ##### preliminaries #####\n",
    "\n",
    "            ##### Graph Generation #####\n",
    "                # undirected graph\n",
    "                nw = SimpleGraph(gene_num)\n",
    "                for i in 1:nrow(cnctdf)\n",
    "                    add_edge!(nw, cnctdf.i1[i], cnctdf.i2[i])\n",
    "                end\n",
    "                # weighted degree by edge value\n",
    "                score = vec(sum(abs.(Matrix(df[Q1, Q1][Q2, Q2])), dims=2))\n",
    "            ##### Graph Generation #####\n",
    "\n",
    "            return nw, pos[Q1, :][Q2, :], cnctdf, clust[Q1][Q2], score\n",
    "        end\n",
    "        export SetNetwork\n",
    "        \"\"\"\n",
    "        fn: figure save name\n",
    "        nw: network graph (one of return value of SetNetwork())\n",
    "        pos: node position (one of return value of SetNetwork())\n",
    "        cnctdf: converted correlation matrix (one of return value of SetNetwork())\n",
    "        clust: cluster number of each gene in network (one of return value of SetNetwork())\n",
    "        k: number of clusters\n",
    "        node_scores: weight of node\n",
    "        node_labels: label of node\n",
    "        node_scaler: multiple for node diameter adjustment (Default: 100)\n",
    "        edge_mode: mode of edges to be drawn\n",
    "                :ALL -> All edges are drawn (Default)\n",
    "                :N -> Only draw edges with negative values.\n",
    "                :P -> Only draw edges with positive values.\n",
    "        edge_threshold: Threshold value of edges to be drawn. (Default: 0.0)\n",
    "        edge_scaler: multiple for edge thickness adjustment (Default: 5)\n",
    "        x_size, y_size: Size of the drawing area (Default: 100, 100)\n",
    "        \"\"\"\n",
    "        function DrawNetwork(fn::String, nw::SimpleGraph, pos::Matrix, cnctdf::DataFrame, clust::Vector{Int64}, k::Int64; node_scores::Vector{}=[], node_labels::Vector{}=[], node_color::Vector=[], node_scaler::Int64=100, edge_mode::Symbol=:ALL, edge_threshold::Float64=0.0, edge_scaler::Int64=5, x_size::Int64=50, y_size::Int64=50)\n",
    "            gene_num = nv(nw)\n",
    "        \n",
    "            # node設計\n",
    "            if length(node_scores) == 0\n",
    "                node_scores = repeat([1], gene_num)\n",
    "            end\n",
    "            node_sizes = node_scores .* node_scaler\n",
    "            if length(node_color) == 0\n",
    "                clust_num = length(unique(vcat(cnctdf.m1, cnctdf.m2)))\n",
    "                color_list = range(LCHuv(65,100,15), stop=LCHuv(65,100,375), length=k)\n",
    "\n",
    "                node_color = [color_list[i] for i in clust]\n",
    "            end\n",
    "            if length(node_labels) == 0\n",
    "                node_labels = repeat([\"\"], gene_num)\n",
    "            end\n",
    "            \n",
    "            # edge設計\n",
    "            edge_colors = RGBA.(1.0, 0.3, 0.0, cnctdf.cor)\n",
    "            edge_colors[cnctdf.cor .< 0.0] = RGBA.(0.0, 0.35, 1.0, abs.(cnctdf.cor[cnctdf.cor .< 0.0]))\n",
    "            Q = (abs.(cnctdf.cor) .< edge_threshold)\n",
    "            if edge_mode == :P\n",
    "                Q = (cnctdf.cor .< 1*edge_threshold)\n",
    "            elseif edge_mode == :N\n",
    "                Q = (cnctdf.cor .> -1*edge_threshold)\n",
    "            end\n",
    "            if sum(Q) == nrow(cnctdf)\n",
    "                println(\"NoEdge\")\n",
    "                return nothing\n",
    "            else\n",
    "                edge_colors[Q] .= RGBA(0.0, 0.0, 0.0, 0.0)\n",
    "            end\n",
    "            edge_sizes = edge_scaler .* abs.(cnctdf.cor)\n",
    "\n",
    "            fig = gplot(\n",
    "                        nw, pos[:, 1] .+ minimum(pos[:, 1]), pos[:, 2] .+ minimum(pos[:, 2]),\n",
    "                        nodesize=node_sizes, nodefillc=node_color, nodelabel=node_labels,\n",
    "                        edgestrokec=edge_colors, edgelinewidth=edge_sizes\n",
    "                    )\n",
    "            Compose.draw(PNG(fn, x_size*cm, y_size*cm), fig)\n",
    "\n",
    "            return nothing\n",
    "        end\n",
    "        export DrawNetwork\n",
    "    ##### draw network #####\n",
    "\n",
    "    ##### Correlation of Phenomenon and Modules #####\n",
    "        function DrawHeatMap(Ax::Axis, X::Vector{Int64}, Y::Vector{Int64}, Mat::Matrix{Float64}, Cnt::Matrix{Int64}, FontSize::Int64)\n",
    "            heatmap!(Ax, X, Y, Mat, colormap=:bwr, colorrange=(-1.0,1.0))\n",
    "            if (FontSize > 0)\n",
    "                for i in X\n",
    "                    for j in Y\n",
    "                        txt = string(round(Mat[i,j],RoundNearestTiesAway,digits=3)) * \" (n=\" * string(Cnt[j]) * \")\"\n",
    "                        text!(Ax, position=(X[i],Y[j]), txt, fontsize=5)\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "            return nothing\n",
    "        end\n",
    "        \"\"\"\n",
    "        argument\n",
    "        df1: dataframe of gene expression\n",
    "        df2: dataframe of Phenomenon\n",
    "        clust: cluster number of each gene (one of return value of SpectralClustering())\n",
    "        fn: fig save name\n",
    "        cor_mode: all gene average (:A_AVG) / positive correlation gene average (:P_AVG) / negative correlation gene average (:N_AVG) / All three types are drawn (:ALL; default)\n",
    "        FontSize: set when drawing the average value of the correlation coefficient on the heat map\n",
    "        \"\"\"\n",
    "        function CorPhenMod(df1::DataFrame, df2::DataFrame, clust::Vector{Int64}, fn::String; cor_mode::Symbol=:ALL, FontSize::Int64=0)\n",
    "            kmax = length(unique(clust))\n",
    "\n",
    "            CorList = []\n",
    "            for i in 1:ncol(df2)\n",
    "                push!(CorList, [[] for i=1:kmax])\n",
    "            end\n",
    "            for k in 1:kmax\n",
    "                for i in 1:ncol(df2)\n",
    "                    buf = df1[clust .== k, :]\n",
    "\t\t            Result_each = zeros(nrow(buf))\n",
    "                    for j in 1:nrow(buf)\n",
    "                        Result_each[j] = cor(Array(buf[j,:]), df2[:,i])\n",
    "                    end\n",
    "\t\t            CorList[i][k] = Result_each\n",
    "                end\n",
    "            end\n",
    "\n",
    "            x = collect(1:ncol(df2))\n",
    "            y = collect(kmax:-1:1)\n",
    "\n",
    "            f = Figure(size=(ncol(df2)*400+200, kmax*30))\n",
    "            ax = []\n",
    "            for i in 1:length(CorList)\n",
    "                push!(ax, Axis(f[1, i], xgridvisible=false, ygridvisible=false, xticksvisible=false, yticksvisible=false, xticks=collect(-1.0:0.5:1.0)))\n",
    "                if (i==1)\n",
    "                    ax[1].yticks = (y,\"module \".*string.(collect(1:kmax)))\n",
    "                else\n",
    "                    ax[i].yticklabelsvisible = false\n",
    "                    linkyaxes!(ax[1], ax[i])\n",
    "                end\n",
    "                hidespines!(ax[i])\n",
    "                ax[i].title = names(df2)[i]\n",
    "                for k in 1:length(CorList[1])\n",
    "                    density!(ax[i], convert.(Float64,CorList[i][k]), bins=100, scale_to=0.1, offset=(kmax-k+1), color=:x, colormap=(:bwr,0.4), colorrange=(-1.0,1.0), strokewidth=1, strokecolor=:black)\n",
    "                end\n",
    "            end\n",
    "            save(fn, f)\n",
    "        end\n",
    "        export CorPhenMod\n",
    "    ##### Correlation of Phenomenon and Modules #####\n",
    "end\n",
    "\n",
    "using CSV, DataFrames\n",
    "using RCall\n",
    "using JLD2\n",
    "using .SGCRNAs\n",
    "\n",
    "## 読込 ##\n",
    "#  行が遺伝子、列が試料；将来的にはWGCNAと同様に行が試料、列が遺伝子にするかもしれない\n",
    "Data = CSV.read(\"Result/Norm/normalizedCounts_coding.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "# 相関係数と傾きを計算。fn以降はオプション引数\n",
    "# fn：指定すると結果を保存する。\n",
    "# mode：前処理の仕方\n",
    "#       :NONE -> 初期値。測定誤差を考慮しない。DEGに対して実行する場合はこれ。\n",
    "#       :LESS -> 標準偏差の分布をプロットしたときにピーク以下を測定誤差として除く。\n",
    "#       :SIGMA -> 標準偏差の分布をプロットしたときにピークとなる値をσとして2σ以内を測定誤差として除く。\n",
    "#       :FTEST -> 標準偏差の分布をプロットしたときにピークとなる値をF検定して測定誤差範囲を決める。\n",
    "# その他のオプション引数に関してはSGCRNA.jl内に記述してあります。\n",
    "CorData, GradData = CCM(Data.Symbol, Matrix(Data[:,5:end]), fn=\"Result/coding-FTEST\", mode=:FTEST);\n",
    "# CorData = load_object(\"Result/coding-FTEST_cor.jld2\");\n",
    "# GradData = load_object(\"Result/coding-FTEST_grad.jld2\");\n",
    "\n",
    "## スペクトラルクラスタリング ##\n",
    "# 重要なオプション引数\n",
    "#       tNodeNum：モジュール内に含まれる最大遺伝子数（初期値：100）この値を超える遺伝子数を有するモジュールのみサブクラスタリングが行われる。\n",
    "#       depthMax：最大何回サブクラスタリングするか。（初期値：５）\n",
    "# その他のオプション引数に関してはSGCRNA.jl内に記述してあります。\n",
    "clust, pos, edge_data = SpectralClustering(CorData, GradData);\n",
    "# save_object(\"Result/coding-FTEST_scdata.jld2\", (clust, pos, edge_data));\n",
    "# (clust, pos, edge_data) = load_object(\"Result/coding-FTEST_scdata.jld2\")\n",
    "\n",
    "## パラメータ設定 ##\n",
    "# ループで全パラメータを実行してもよい\n",
    "# d=3が大体ちょうどいい\n",
    "d = 3; k = maximum(clust[d]);\n",
    "GeneClust = DataFrame(Gene=names(edge_data), Module=clust[d]);\n",
    "GeneClust |> CSV.write(\"Result/GeneCluster_coding.tsv\", delim='\\t', writeheader=true);\n",
    "\n",
    "## ネットワーク設定 ##\n",
    "# 全て描画する場合\n",
    "nw, new_pos, cnctdf, new_clust, score = SetNetwork(edge_data, clust[d], pos, il=collect(1:k));\n",
    "# 特定のモジュールだけ描画したい場合\n",
    "nw, new_pos, cnctdf, new_clust, score = SetNetwork(edge_data, clust[d], pos, il=[2,8]);\n",
    "# save_object(\"Result/coding_nwdata.jld2\", (nw, new_pos, cnctdf, new_clust, score, edge_data))\n",
    "# (nw, new_pos, cnctdf, new_clust, score, edge_data) = load_object(\"Result/coding_nwdata.jld2\"); k = maximum(new_clust);\n",
    "\n",
    "## ノードラベル設定（オプション） ##\n",
    "# 特定の遺伝子名だけ描画したい場合\n",
    "NodeLabel = [x in [\"BARX1\",\"BARX2\",\"COL14A1\",\"GDF5\",\"MMP1\",\"PRG4\",\"COL10A1\",\"IHH\",\"MMP13\",\"PTH1R\",\"RUNX2\",\"SP7\"] ? x : \"\" for x in names(edge_data)];\n",
    "# 特定の遺伝子名だけ描画したい場合(特定のモジュールだけ描画する場合)\n",
    "NodeLabel = unique(sort(vcat(cnctdf.e1,cnctdf.e2)))\n",
    "NodeLabel = [x in [\"BARX1\",\"BARX2\",\"COL14A1\",\"GDF5\",\"MMP1\",\"PRG4\",\"COL10A1\",\"IHH\",\"MMP13\",\"PTH1R\",\"RUNX2\",\"SP7\"] ? x : \"\" for x in NodeLabel]\n",
    "\n",
    "## ネットワーク描画 ##\n",
    "# 全てのエッジを描画\n",
    "DrawNetwork(\"Result/Fig/bulk_AllNetWork_coding-FTEST.png\", nw, new_pos, cnctdf, new_clust, k, node_scores=score, edge_mode=:A)\n",
    "# エッジスコアの絶対値が0.9以上を描画\n",
    "DrawNetwork(\"Result/Fig/bulk_AllNetWork_coding-FTEST-0.9.png\", nw, new_pos, cnctdf, new_clust, k, node_scores=score, edge_mode=:A, edge_threshold=0.9)\n",
    "# エッジスコアが負のみ描画\n",
    "DrawNetwork(\"Result/Fig/bulk_NegativeNetWork_coding-FTEST-0.9.png\", nw, new_pos, cnctdf, new_clust, k, node_scores=score, edge_mode=:N)\n",
    "# エッジスコアが正かつエッジスコアの絶対値が0.9以上のみ描画。指定したノードラベルも描画\n",
    "DrawNetwork(\"Result/Fig/bulk_PositiveNetWork_coding-FTEST-0.9.png\", nw, new_pos, cnctdf, new_clust, k, node_scores=score, node_labels=NodeLabel, edge_mode=:P, edge_threshold=0.9)\n",
    "\n",
    "\n",
    "# GO解析\n",
    "for i in 1:k\n",
    "    commun = names(edge_data)[new_clust .== i]\n",
    "    R\"\"\"\n",
    "        Pkgs <- c(\"clusterProfiler\", \"enrichplot\", \"org.Hs.eg.db\", \"ggplot2\")\n",
    "        for(p in Pkgs) suppressMessages(library(p, character.only=T))\n",
    "\n",
    "        options(ggrepel.max.overlaps = Inf)\n",
    "\n",
    "        Analysis_GO <- function(x) {\n",
    "            result <- enrichGO(\n",
    "                                gene=x,\n",
    "                                keyType=\"SYMBOL\",\n",
    "                                OrgDb=org.Hs.eg.db,\n",
    "                                ont=\"ALL\",                #\"BP\",\"CC\",\"MF\",\"ALL\"\n",
    "                                pAdjustMethod=\"BH\",\n",
    "                                pvalueCutoff=0.05,\n",
    "                                qvalueCutoff=0.05\n",
    "                            )\n",
    "        }\n",
    "        PlotFig <- function(res) {\n",
    "            res_simple <- simplify(res)\n",
    "            Fig <- clusterProfiler::dotplot(res_simple)\n",
    "            ggsave(paste0(\"Result/Fig/bulk_GO_cluser-\",$i,\"_dot_simple.png\"), plot=Fig, dpi=400)\n",
    "            Fig <- clusterProfiler::dotplot(res)\n",
    "            ggsave(paste0(\"Result/Fig/bulk_GO_cluser-\",$i,\"_dot_all.png\"), plot=Fig, dpi=400)\n",
    "        }\n",
    "\n",
    "        result <- try(Analysis_GO($commun), silent=FALSE)\n",
    "        if (class(result) != \"try-error\") {\n",
    "            y <- try(PlotFig(result), silent=FALSE)\n",
    "        }\n",
    "    \"\"\"\n",
    "end\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.9.0",
   "language": "julia",
   "name": "julia-1.9"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
